### PR TITLE
add package: abacus

### PR DESCRIPTION
**Pull request by Bitllion**

### CODE REVIEW

**Comment by Bitllion** [comment]
@tldahlgren @alalazo

**Comment by Bitllion** [comment]
@lhxone

**Comment by Bitllion** [comment]
@lhxone

**Comment by Bitllion** [comment]
@tldahlgren @alalazo @lhxone @chuckatkins  @adamjstewart  
 The CI test has passed. Please approve my merge request

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    homepage = "http://abacus.ustc.edu.cn/"
+
+    version(
+        '2.2.1',
+        '14feca1d8d1ce025d3f263b85ebfbebc1a1efff704b6490e95b07603c55c1d63',
+        url='https://github.com/abacusmodeling/abacus-develop/archive/refs/tags/v2.2.1.tar.gz')
+    version(
+        '2.2.0',
+        '09d4a2508d903121d29813a85791eeb3a905acbe1c5664b8a88903f8eda64b8f',
+        url='https://github.com/abacusmodeling/abacus-develop/archive/refs/tags/v2.2.0.tar.gz')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/abacus/package.py]: The `url` arguments for the two versions are identical except for the version.  Is there a reason not to use it as a property instead?

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    mkl_message = 'Need to set dependent variant to threads=openmp'
+
+    conflicts('+openmp',
+              when='^intel-oneapi-mkl',
+              msg=mkl_message)
+
+    build_directory = 'source'
+
+    def edit(self, spec, prefix):
+        if self.version == Version('2.2.1'):
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/abacus/package.py]: Curious.  Why not use `self.spec.satisfies('@2.2.1')` here and below?

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/abacus/package.py]: For an example, see https://spack.readthedocs.io/en/latest/packaging_guide.html#testing-spec-constraints.

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                'LIBXC_DIR = ' +
+                spec['libxc'].prefix)
+            makefile_vars.filter(
+                '^\\s*CEREAL_DIR\\s*=.*',
+                'CEREAL_DIR = ' +
+                spec['cereal'].prefix)
+            makefile_vars.filter(
+                '^\\s*ELPA_DIR\\s*=.*',
+                'ELPA_DIR = ' +
+                spec['elpa'].prefix)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/abacus/package.py]: There's a lot of redundancy for the statements for this and the next version and could present a maintainability issue in the long run.

For example, these statements are identical.

Could this be refactored?

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+            makefile_system = FileFilter('source/Makefile.system')
+            makefile_system.filter(
+                '^\\s*ELPA_INCLUDE_DIR\\s*=.*',
+                'ELPA_INCLUDE_DIR = -I${ELPA_DIR}/include/' +
+                'elpa_openmp-{0}'.format(
+                    spec['elpa'].version))
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/abacus/package.py]: This section differs from the next by setting and using `makefile_system` and adding `_DIR` to `ELPA_INCLUDE`.

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+            pwd = os.getcwd() + "/source"
+            with open(pwd + '/Makefile.system', "r") as f:
+                flist = f.readlines()
+                flist[22] = 'ELPA_LIB     = -L${ELPA_LIB_DIR} \
+                -lelpa_openmp -Wl,-rpath=${ELPA_LIB_DIR}\n'
+                flist[23] = '#ELPA_LIB     = -L${ELPA_LIB_DIR} \
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/abacus/package.py]: The `flist` indices are the only differences between this and the next section (22 and 23 here versus 25 and 26 for `2.2.0`.

**Review by tldahlgren — COMMENTED** [review]
I have a couple of questions related to the `url` and redundancy in the `edit` method.

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 3 above)
Comment (Bitllion) [reply var/spack/repos/builtin/packages/abacus/package.py]: OK, I took your suggestion and updated abacus's "package.py"

File: var/spack/repos/builtin/packages/abacus/package.py (hunk 1 above)
Comment (Bitllion) [reply var/spack/repos/builtin/packages/abacus/package.py]: OK, I took your suggestion and updated abacus's "package.py"

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    depends_on("elpa+openmp")
+    depends_on("cereal")
+    depends_on("libxc")
+    depends_on("fftw")
+    depends_on("mpi", when="+mpi", type=("build", "link", "run"))
+    depends_on("mkl")
+
+    mkl_message = "Need to set dependent variant to threads=openmp"
+
+    conflicts("+openmp", msg=mkl_message)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/abacus/package.py]: I find this a bit confusing.  You're setting the default to `+openmp` then declare a conflict when it is set.

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    )
+
+    depends_on("elpa+openmp")
+    depends_on("cereal")
+    depends_on("libxc")
+    depends_on("fftw")
+    depends_on("mpi", when="+mpi", type=("build", "link", "run"))
+    depends_on("mkl")
+
+    mkl_message = "Need to set dependent variant to threads=openmp"
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/abacus/package.py]: If you want to add `threads` as a multi-valued variant, a simple example can be found at https://github.com/spack/spack/blob/17e71a675aed0695a27f3127291bd6028f40d3f9/var/spack/repos/builtin/packages/armpl/package.py#L104.

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        sha256="14feca1d8d1ce025d3f263b85ebfbebc1a1efff704b6490e95b07603c55c1d63",
+    )
+    version(
+        "2.2.0",
+        sha256="09d4a2508d903121d29813a85791eeb3a905acbe1c5664b8a88903f8eda64b8f",
+    )
+
+    variant("mpi", default=True, description="Builds with MPI support. Requires MPI2+")
+    variant(
+        "openmp", default=True, description="Enables OpenMP threads. Use threaded FFTW3"
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/abacus/package.py]: FWIW. There's an example of specifying `fftw3` as a variant at https://github.com/spack/spack/blob/17e71a675aed0695a27f3127291bd6028f40d3f9/var/spack/repos/builtin/packages/cdo/package.py#L58

with the dependency at https://github.com/spack/spack/blob/17e71a675aed0695a27f3127291bd6028f40d3f9/var/spack/repos/builtin/packages/cdo/package.py#L83.

File: var/spack/repos/builtin/packages/abacus/package.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    )
+
+    depends_on("elpa+openmp")
+    depends_on("cereal")
+    depends_on("libxc")
+    depends_on("fftw")
+    depends_on("mpi", when="+mpi", type=("build", "link", "run"))
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/abacus/package.py]: Given the comment for the variant, I *think* what you want is something along the lines:

```suggestion
    depends_on("mpi@2:", when="+mpi", type=("build", "link", "run"))
```

Comment (Bitllion) [reply var/spack/repos/builtin/packages/abacus/package.py]: Sorry, I don't quite understand the meaning of spack's grammar. Thank you for your suggestions.
Abacus's MPI dependency is necessary, so I deleted the MPI parameters "when" and "variant".
The variable OpenMP is related to the dependent elpa. I modified the logical judgment of elpa when using OpenMP.

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**